ansible/inventory/multi_inventory.py

```diff
@@ -119,144 +227,103 @@ def load_yaml_config(self, conf_file=None):
-           and call self.get_provider_tags.
-        '''
+    def run_account_providers(self, force=False, valid_cache=True, account=None):
+        """Setup the provider call with proper variables and call self.get_provider_tags"""
         try:
-            all_results = []
-            tmp_dir_paths = []
+            # if account parameter is present we only want refresh that specific account
+            tmp_accounts = self.accounts
+            if account:
+                tmp_accounts = [acc for acc in self.accounts if acc.name == account]
+
             processes = {}
-            for acc_name, account in self.config['accounts'].items():
-                tmp_dir = None
-                if account.has_key('provider_files'):
-                    tmp_dir = MultiInventory.generate_config(account['provider_files'])
-                    tmp_dir_paths.append(tmp_dir)
-
-                # Update env vars after creating provider_config_files
-                # so that we can grab the tmp_dir if it exists
-                env = account.get('env_vars', {})
-                if env and tmp_dir:
-                    for key, value in env.items():
-                        env[key] = Template(value).substitute(tmpdir=tmp_dir)
-
-                provider = account['provider']
-                processes[acc_name] = self.get_provider_tags(provider, env)
+            for acc in tmp_accounts:
+                if self.args.get('debug', False):
+                    print "Calling run_provider for account: %s" % acc.name
+                processes[acc.name] = acc.run_provider(force=force, valid_cache=valid_cache)
 
             # for each process collect stdout when its available
-            for name, process in processes.items():
-                out, err = process.communicate()
-                all_results.append({
-                    "name": name,
-                    "out": out.strip(),
-                    "err": err.strip(),
-                    "code": process.returncode
-                })
-
-        finally:
-            # Clean up the mkdtemp dirs
-            for tmp_dir in tmp_dir_paths:
-                shutil.rmtree(tmp_dir)
-
-        return all_results
+            for acc in tmp_accounts:
+                if not processes[acc.name]['cached']:
+                    proc = processes[acc.name]['proc']
+                    out, err = proc.communicate()
+
+                    # For any non-zero, raise an error on it
+                    if proc.returncode != 0:
+                        err_msg = ['\nProblem fetching account: %s' % acc.name,
+                                   'Error Code: %s' % proc.returncode,
+                                   'StdErr: %s' % err,
+                                   'Stdout: %s' % out,
+                                  ]
+                        raise RuntimeError('\n'.join(err_msg))
+
+                    acc.account_data = json.loads(out)
+                    acc.write_to_cache()
```

Comment:
This section feels like it should be part of the account class.

Is there a reason to run it outside of the account class?